Sources/Controllers/Map/Layers/OAMyPositionLayer.mm

```diff
@@ -469,83 +534,194 @@ - (void) generateMarkersCollection
         locationAndCourseMarkerBuilder.setAccuracyCircleBaseColor(OsmAnd::ColorRGB(0x20, 0xad, 0xe5));
         locationAndCourseMarkerBuilder.setBaseOrder(baseOrder--);
         locationAndCourseMarkerBuilder.setIsHidden(true);
+        locationAndCourseMarkerBuilder.setModel3DMaxSizeInPixels(int(MODEL_3D_MAX_SIZE_DP * _textScaleFactor * [[UIScreen mainScreen] scale]));
         
-        OALocationIcon *locIcon = [OALocationIcon withLocationIcon:mode.getLocationIcon];
-        OANavigationIcon *navIcon = [OANavigationIcon withNavigationIcon:mode.getNavigationIcon];
         UIColor *iconColor = UIColorFromRGB(mode.getIconColor);
         
+        NSString *locationIconName = mode.getLocationIcon;
+        NSString *navigationIconName = mode.getNavigationIcon;
+        sk_sp<SkImage> navigationSkImage;
+        sk_sp<SkImage> navigationLostSkImage;
+        sk_sp<SkImage> locationSkImage;
+        sk_sp<SkImage> locationHeadingSkImage;
+        
+        OAModel3dWrapper *navigationModel;
+        OAModel3dWrapper *locationModel;
+        std::shared_ptr<const OsmAnd::Model3D> navigationModelCpp;
+        std::shared_ptr<const OsmAnd::Model3D> locationModelCpp;
+        
+        BOOL notInited = NO;
+        
+        OANavigationIcon *navIcon = [OANavigationIcon withIconName:navigationIconName];
+        if ([navIcon isModel])
+        {
+            navigationModel = [Model3dHelper.shared getModelWithModelName:navigationIconName callbackOnLoad:^(OAModel3dWrapper * _Nullable model) {
+                [weakSelf refreshMarkersCollection];
```

Comment:
Refresh all markers again on every model load is bad thing. You may load all models right away, but generateMarkersCollection should not be called on every load.